Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp

```diff
@@ -663,6 +668,12 @@ ExceptionOr<void> ApplePayPaymentHandler::shippingAddressUpdated(Vector<RefPtr<A
     auto modifierException = firstApplicableModifier();
     if (modifierException.hasException())
         return modifierException.releaseException();
+
+    // firstApplicableModifier() may run JS, which may abort the request so we need
+    // to make sure we still have an active session.
+    if (!paymentCoordinator().hasActiveSession())
```

Comment:
is this one actually a problem?  IIRC we serialize the `modifiers` to a string when first ingesting theme, so all `firstApplicableModifier()` does is parse it back into a JSON object and then convert that to the relevant IDL/C++ dictionary (i.e. there shouldn't be any opportunity for user JS to run when doing this since it's JSON parsed from a string)

if it is an issue, should we be doing this check for all the other places that `firstApplicableModifier()` is used (e.g. `ApplePayPaymentHandler::shippingOptionUpdated`)?

Reply:
Yes, this one definitely was needed to address the crashes. I believe it is this call inside `firstApplicableModifier()` that can run script:
```
auto applePayModifier = convertDictionary<ApplePayModifier>(lexicalGlobalObject, WTFMove(data));
```

I'll look at other call sites for firstApplicableModifier(). They likely need post validation too.

Reply:
I added checks after the other `firstApplicableModifier()` call sites.

Reply:
right but the `data` is coming from a `JSONParse` from a string

```
data = JSONParse(&lexicalGlobalObject, serializedModifierData[i]);
```

so there shouldn't be any way for user JS to run from that.  the only issue i can see is if the given JSON doesnt actually match `ApplePayModifier`, but that should already throw an exception, so i dont see how we can successfully return from `ApplePayPaymentHandler::firstApplicableModifier` and yet somehow also abort.  can you elaborate on the scenario where this is crashing?

Reply:
You're right, I'll double check and report back. Maybe I can move the hasActiveSession() check earlier.

Reply:
It was computeErrors() at the call site which could execute JavaScript so I'm moving the check after computeErrors() calls instead.